### PR TITLE
Use timeseries instead of topN for query with breakout by timestamp

### PR DESCRIPTION
**Pull request by moumny**
I did a quick benchmark with 100 queries it takes 12s using topn and 1.8s using timeseries.

New to the project and clojure so please take good look at it.

Still need to test order-by once #2859 is fixed.

### CODE REVIEW

File: src/metabase/driver/druid/query_processor.clj [hunk 1]
```diff
@@ -14,17 +14,21 @@
...
-(derive ::grouped-ag-query ::ag-query)
-(derive ::topN             ::grouped-ag-query)
-(derive ::groupBy          ::grouped-ag-query)
+
+;;             +-----> ::select      +----> :groupBy
+;; ::query ----|                     |
+;;             +----> ::ag-query ----+----> ::topN
+;;                                   |                       +----> total
+;;                                   +----> ::timeseries ----|
+;;                                                           +----> grouped-timeseries
```
Comment (camsaul) [inline src/metabase/driver/druid/query_processor.clj]: what's the reasoning behind removing `::grouped-ag-query`?

File: src/metabase/driver/druid/query_processor.clj [hunk 2]
```diff
@@ -214,6 +221,21 @@
                                  "}")
     :year            (extract:timeFormat "yyyy")))
 
+(defn- unit->granularity
+  [unit]
+  (let [iso-period (case unit
+                  :minute  "PT1M"
```
Comment (camsaul) [inline src/metabase/driver/druid/query_processor.clj]: looks like the indentation is off here.

```clojure
(case unit
  :minute ...)
```

File: src/metabase/driver/druid/query_processor.clj [hunk 3]
```diff
@@ -492,10 +530,14 @@
...
 
 (defmethod post-process ::select     [_ results] (->> results first :result :events (map :event)))
-(defmethod post-process ::timeseries [_ results] (map :result results))
+(defmethod post-process ::total [_ results] (map :result results))
 (defmethod post-process ::topN       [_ results] (-> results first :result))
 (defmethod post-process ::groupBy    [_ results] (map :event results))
 
+(defmethod post-process ::grouped-timeseries [_ results]
+  (map (fn [event]
+    (conj {:timestamp (:timestamp event)} (:result event))) results))
```
Comment (camsaul) [inline src/metabase/driver/druid/query_processor.clj]: You can just use `for` here.

```clojure
(defmethod post-process ::grouped-timeseries [_ results]
  (for [event results]
    (conj {:timestamp (:timestamp event)} (:result event))))
```

**Comment by camsaul** [comment]
@moumny thanks for this PR, I will give it a try

**Comment by camsaul** [comment]
Any reason why it's "WIP"?

File: src/metabase/driver/druid/query_processor.clj (hunk 3 above)
Comment (camsaul) [inline src/metabase/driver/druid/query_processor.clj]: What's being `conj`ed here?

**Comment by camsaul** [comment]
What's the reason you called an ungrouped timeseries "total"?

**Comment by camsaul** [comment]
@moumny I'm seeing several failures when I run tests against this branch. Looks like it breaks date bucketing and formatting in a few places (results edited for brevity):

```clojure
$ ENGINES=druid lein test metabase.timeseries-query-processor-test

failure in (timeseries_query_processor_test.clj:389) : metabase.timeseries-query-processor-test
(expect (metabase.test.data.datasets/do-with-engine :druid expected-1699996181) (metabase.test.data.datasets/do-with-engine :druid actual-296160825))

           
expected:
 {:columns ["timestamp" "count"], :rows [["2013-01-03+0000" 1] ["2013-01-10+0000" 1] ["2013-01-19+0000" 1] ["2013-01-22+0000" 1] ["2013-01-23+0000" 1]]}
 
was:
 {:columns ["timestamp" "count"],
 :rows
 [("2013-01-03T00:00:00.000Z" 1)
  ("2013-01-10T00:00:00.000Z" 1)
  ("2013-01-19T00:00:00.000Z" 1)
  ("2013-01-22T00:00:00.000Z" 1)
  ("2013-01-23T00:00:00.000Z" 1)
  ("2013-01-25T00:00:00.000Z" 1)
  ...
  ("2015-12-26T00:00:00.000Z" 2)
  ("2015-12-29T00:00:00.000Z" 1)]}


           
in expected, not actual:
{:rows [["2013-01-03+0000"] ["2013-01-10+0000"] ["2013-01-19+0000"] ["2013-01-22+0000"] ["2013-01-23+0000"]]}

           
in actual, not expected:
{:rows
 [["2013-01-03T00:00:00.000Z"]
  ["2013-01-10T00:00:00.000Z"]
  ["2013-01-19T00:00:00.000Z"]
  ["2013-01-22T00:00:00.000Z"]
  ["2013-01-23T00:00:00.000Z"]
  ("2013-01-25T00:00:00.000Z" 1)
  ("2013-01-26T00:00:00.000Z" 1)
  ("2013-01-27T00:00:00.000Z" 1)
  ("2013-02-06T00:00:00.000Z" 2)
  ...
  ("2015-12-26T00:00:00.000Z" 2)
  ("2015-12-29T00:00:00.000Z" 1)]}


failure in (timeseries_query_processor_test.clj:402) : metabase.timeseries-query-processor-test
(expect (metabase.test.data.datasets/do-with-engine :druid expected--1787425538) (metabase.test.data.datasets/do-with-engine :druid actual--1393814129))

           
expected:
 {:columns ["timestamp" "count"], :rows [["2013-01-03T08:00:00+0000" 1] ["2013-01-10T08:00:00+0000" 1] ["2013-01-19T08:00:00+0000" 1] ["2013-01-22T08:00:00+0000" 1] ["2013-01-23T08:00:00+0000" 1]]}
 
was:
 {:columns ["timestamp" "count"],
 :rows
 [("2013-01-03T08:00:00.000Z" 1)
  ("2013-01-10T08:00:00.000Z" 1)
  ("2013-01-19T08:00:00.000Z" 1)
  ...
  ("2015-12-26T08:00:00.000Z" 2)
  ("2015-12-29T08:00:00.000Z" 1)]}


           
in expected, not actual:
{:rows [["2013-01-03T08:00:00+0000"] ["2013-01-10T08:00:00+0000"] ["2013-01-19T08:00:00+0000"] ["2013-01-22T08:00:00+0000"] ["2013-01-23T08:00:00+0000"]]}

           
in actual, not expected:
{:rows
 [["2013-01-03T08:00:00.000Z"]
  ["2013-01-10T08:00:00.000Z"]
  ["2013-01-19T08:00:00.000Z"]
  ["2013-01-22T08:00:00.000Z"]
  ["2013-01-23T08:00:00.000Z"]
  ("2013-01-25T08:00:00.000Z" 1)
  ("2013-01-26T08:00:00.000Z" 1)
  ...
  ("2015-12-26T08:00:00.000Z" 2)
  ("2015-12-29T08:00:00.000Z" 1)]}


failure in (timeseries_query_processor_test.clj:424) : metabase.timeseries-query-processor-test
(expect (metabase.test.data.datasets/do-with-engine :druid expected--1787425538) (metabase.test.data.datasets/do-with-engine :druid actual-1271342367))

           
expected:
 {:columns ["timestamp" "count"], :rows [["2013-01-03T08:00:00+0000" 1] ["2013-01-10T08:00:00+0000" 1] ["2013-01-19T08:00:00+0000" 1] ["2013-01-22T08:00:00+0000" 1] ["2013-01-23T08:00:00+0000" 1]]}
 
was:
 {:columns ["timestamp" "count"],
 :rows
 [("2013-01-03T08:00:00.000Z" 1)
  ("2013-01-10T08:00:00.000Z" 1)
  ("2013-01-19T08:00:00.000Z" 1)
  ...
  ("2015-12-26T08:00:00.000Z" 2)
  ("2015-12-29T08:00:00.000Z" 1)]}


           
in expected, not actual:
{:rows [["2013-01-03T08:00:00+0000"] ["2013-01-10T08:00:00+0000"] ["2013-01-19T08:00:00+0000"] ["2013-01-22T08:00:00+0000"] ["2013-01-23T08:00:00+0000"]]}

           
in actual, not expected:
{:rows
 [["2013-01-03T08:00:00.000Z"]
  ["2013-01-10T08:00:00.000Z"]
  ["2013-01-19T08:00:00.000Z"]
  ["2013-01-22T08:00:00.000Z"]
  ["2013-01-23T08:00:00.000Z"]
  ("2013-01-25T08:00:00.000Z" 1)
  ("2013-01-26T08:00:00.000Z" 1)
  ("2013-01-27T08:00:00.000Z" 1)
  ...
  ("2015-12-26T08:00:00.000Z" 2)
  ("2015-12-29T08:00:00.000Z" 1)]}


failure in (timeseries_query_processor_test.clj:447) : metabase.timeseries-query-processor-test
(expect (metabase.test.data.datasets/do-with-engine :druid expected-63907517) (metabase.test.data.datasets/do-with-engine :druid actual-617099742))

           
expected:
 {:columns ["timestamp" "count"], :rows [["2012-12-30" 1] ["2013-01-06" 1] ["2013-01-13" 1] ["2013-01-20" 4] ["2013-01-27" 1]]}
 
was:
 {:columns ["timestamp" "count"],
 :rows
 [("2012-12-31T00:00:00.000Z" 1)
  ("2013-01-07T00:00:00.000Z" 1)
  ("2013-01-14T00:00:00.000Z" 1)
  ("2013-01-21T00:00:00.000Z" 5)
  ("2013-02-04T00:00:00.000Z" 2)
  ...
  ("2015-12-21T00:00:00.000Z" 4)
  ("2015-12-28T00:00:00.000Z" 1)]}


           
in expected, not actual:
{:rows [["2012-12-30"] ["2013-01-06"] ["2013-01-13"] ["2013-01-20" 4] ["2013-01-27" 1]]}

           
in actual, not expected:
{:rows
 [["2012-12-31T00:00:00.000Z"]
  ["2013-01-07T00:00:00.000Z"]
  ["2013-01-14T00:00:00.000Z"]
  ["2013-01-21T00:00:00.000Z" 5]
  ["2013-02-04T00:00:00.000Z" 2]
  ("2013-02-11T00:00:00.000Z" 2)
  ...
  ("2015-12-21T00:00:00.000Z" 4)
  ("2015-12-28T00:00:00.000Z" 1)]}


failure in (timeseries_query_processor_test.clj:460) : metabase.timeseries-query-processor-test
(expect (metabase.test.data.datasets/do-with-engine :druid expected-1699996181) (metabase.test.data.datasets/do-with-engine :druid actual--1898165403))

           
expected:
 {:columns ["timestamp" "count"], :rows [["2013-01-03+0000" 1] ["2013-01-10+0000" 1] ["2013-01-19+0000" 1] ["2013-01-22+0000" 1] ["2013-01-23+0000" 1]]}
 
was:
 {:columns ["timestamp" "count"],
 :rows
 [("2013-01-03T00:00:00.000Z" 1)
  ("2013-01-10T00:00:00.000Z" 1)
  ("2013-01-19T00:00:00.000Z" 1)
  ...
  ("2015-12-26T00:00:00.000Z" 2)
  ("2015-12-29T00:00:00.000Z" 1)]}


           
in expected, not actual:
{:rows [["2013-01-03+0000"] ["2013-01-10+0000"] ["2013-01-19+0000"] ["2013-01-22+0000"] ["2013-01-23+0000"]]}

           
in actual, not expected:
{:rows
 [["2013-01-03T00:00:00.000Z"]
  ["2013-01-10T00:00:00.000Z"]
  ["2013-01-19T00:00:00.000Z"]
  ["2013-01-22T00:00:00.000Z"]
  ["2013-01-23T00:00:00.000Z"]
  ("2013-01-25T00:00:00.000Z" 1)
  ("2013-01-26T00:00:00.000Z" 1)
  ("2013-01-27T00:00:00.000Z" 1)
  ("2013-02-06T00:00:00.000Z" 2)
  ...
  ("2015-12-26T00:00:00.000Z" 2)
  ("2015-12-29T00:00:00.000Z" 1)]}


failure in (timeseries_query_processor_test.clj:525) : metabase.timeseries-query-processor-test
(expect (metabase.test.data.datasets/do-with-engine :druid expected-1046461676) (metabase.test.data.datasets/do-with-engine :druid actual--1811045274))

           
expected:
 {:columns ["timestamp" "count"], :rows [["2013-01-01" 8] ["2013-02-01" 11] ["2013-03-01" 21] ["2013-04-01" 26] ["2013-05-01" 23]]}
 
was:
 {:columns ["timestamp" "count"],
 :rows
 [("2013-01-01T00:00:00.000Z" 8)
  ("2013-02-01T00:00:00.000Z" 11)
  ("2013-03-01T00:00:00.000Z" 21)
  ("2013-04-01T00:00:00.000Z" 26)
  ...
  ("2015-11-01T00:00:00.000Z" 24)
  ("2015-12-01T00:00:00.000Z" 13)]}


           
in expected, not actual:
{:rows [["2013-01-01"] ["2013-02-01"] ["2013-03-01"] ["2013-04-01"] ["2013-05-01"]]}

           
in actual, not expected:
{:rows
 [["2013-01-01T00:00:00.000Z"]
  ["2013-02-01T00:00:00.000Z"]
  ["2013-03-01T00:00:00.000Z"]
  ["2013-04-01T00:00:00.000Z"]
  ["2013-05-01T00:00:00.000Z"]
  ("2013-06-01T00:00:00.000Z" 26)
  ("2013-07-01T00:00:00.000Z" 20)
  ("2013-08-01T00:00:00.000Z" 22)
  ...
  ("2015-11-01T00:00:00.000Z" 24)
  ("2015-12-01T00:00:00.000Z" 13)]}


failure in (timeseries_query_processor_test.clj:551) : metabase.timeseries-query-processor-test
(expect (metabase.test.data.datasets/do-with-engine :druid expected--1970927549) (metabase.test.data.datasets/do-with-engine :druid actual-139692756))

           
expected:
 {:columns ["timestamp" "count"], :rows [["2013-01-01" 40] ["2013-04-01" 75] ["2013-07-01" 55] ["2013-10-01" 65] ["2014-01-01" 107]]}
 
was:
 {:columns ["timestamp" "count"],
 :rows
 [("2013-01-01T00:00:00.000Z" 40)
  ("2013-04-01T00:00:00.000Z" 75)
  ("2013-07-01T00:00:00.000Z" 55)
  ("2013-10-01T00:00:00.000Z" 65)
  ...
  ("2015-07-01T00:00:00.000Z" 85)
  ("2015-10-01T00:00:00.000Z" 58)]}


           
in expected, not actual:
{:rows [["2013-01-01"] ["2013-04-01"] ["2013-07-01"] ["2013-10-01"] ["2014-01-01"]]}

           
in actual, not expected:
{:rows
 [["2013-01-01T00:00:00.000Z"]
  ["2013-04-01T00:00:00.000Z"]
  ["2013-07-01T00:00:00.000Z"]
  ["2013-10-01T00:00:00.000Z"]
  ["2014-01-01T00:00:00.000Z"]
  ("2014-04-01T00:00:00.000Z" 138)
  ("2014-07-01T00:00:00.000Z" 138)
  ("2014-10-01T00:00:00.000Z" 115)
  ("2015-01-01T00:00:00.000Z" 53)
  ("2015-04-01T00:00:00.000Z" 71)
  ("2015-07-01T00:00:00.000Z" 85)
  ("2015-10-01T00:00:00.000Z" 58)]}


failure in (timeseries_query_processor_test.clj:576) : metabase.timeseries-query-processor-test
(expect (metabase.test.data.datasets/do-with-engine :druid expected--479702787) (metabase.test.data.datasets/do-with-engine :druid actual--1956682366))

           
expected:
 {:columns ["timestamp" "count"], :rows [["2013" 235] ["2014" 498] ["2015" 267]]}
 
was:
 {:columns ["timestamp" "count"], :rows [("2013-01-01T00:00:00.000Z" 235) ("2014-01-01T00:00:00.000Z" 498) ("2015-01-01T00:00:00.000Z" 267)]}


           
in expected, not actual:
{:rows [["2013"] ["2014"] ["2015"]]}

           
in actual, not expected:
{:rows [["2013-01-01T00:00:00.000Z"] ["2014-01-01T00:00:00.000Z"] ["2015-01-01T00:00:00.000Z"]]}

Ran 81 tests containing 81 assertions in 23063 msecs
8 failures, 0 errors.
```

**Comment by moumny** [comment]
Thanks for taking a look.
The tests were failing because LIMIT was disabled for the timeseries. In PlyQL (Druid sql parser) they do the limit in the client side. Im not sure this is ideal, so now we fallback to TopN if you limit.

Total is common name for timeseries of size 1.

**Comment by camsaul** [comment]
@moumny this looks good to me. Thanks for the contribution

**Pull request merged**